atomic_reactor/plugins/pre_bump_release.py

```diff
@@ -300,25 +274,74 @@ def run(self):
-                                      dockerfile_labels)
+            if not is_scratch_build(self.workflow):
+                self.check_build_existence_for_explicit_release(component, version, release)
+        else:
+            # release not set or release should be appended
+            release = self.next_release_general(component, version, original_release)
 
         if self.reserve_build and not is_scratch_build(self.workflow):
-            self.reserve_build_in_koji(component, version, release, release_label,
-                                       dockerfile_labels)
+            release = self.reserve_build_in_koji(
+                component, version, release, explicitly_provided_release=bool(original_release)
+            )
+
+        # Always set preferred release label - other will be set if old-style label is present
+        preferred_release_label = Labels.LABEL_NAMES[Labels.LABEL_TYPE_RELEASE][0]
+
+        if dockerfile_labels.get(preferred_release_label) != release:
+            self.log.info("setting %s=%s", preferred_release_label, release)
+            # Write the label back to the file (this is a property setter)
+            dockerfile_labels[preferred_release_label] = release
+
+    def _get_nvr(self, dockerfile_labels) -> Tuple[str, str, Optional[str]]:
+        """Get the component, version and release labels from the Dockerfile."""
+        labels = Labels(dockerfile_labels)
+
+        component_label = labels.get_name(Labels.LABEL_TYPE_COMPONENT)
+        component: Optional[str] = dockerfile_labels.get(component_label)
+
+        version_label = labels.get_name(Labels.LABEL_TYPE_VERSION)
+        version: Optional[str] = dockerfile_labels.get(version_label)
+
+        release_label = labels.get_name(Labels.LABEL_TYPE_RELEASE)
+        release: Optional[str] = dockerfile_labels.get(release_label)
+
+        missing_labels = {}
+
+        # component: missing X, empty OK
+        if component is None:
+            self.log.error("missing label: %s", component_label)
+            missing_labels[component_label] = "missing"
+
+        # version: missing X, empty X
+        if version is None:
+            self.log.error("missing label: %s", version_label)
+            missing_labels[version_label] = "missing"
+        elif not version:
+            self.log.error("empty label: %s", version_label)
+            missing_labels[version_label] = "empty"
+
+        # release: missing OK, (present but) empty X
+        if (release is not None) and not release:
+            self.log.error("empty label: %s", release_label)
+            missing_labels[release_label] = "empty"
+
+        if missing_labels:
+            raise RuntimeError(
+                "Required labels are missing or empty or using undefined variables: {}"
+                .format(missing_labels)
+            )
+
+        # For type-checkers: narrow the type of component and version to str
+        assert component is not None and version is not None
```

Comment:
do we really now need to add asserts in our code??

Reply:
If the code looked like this, we wouldn't:
```python
if not component:
    raise RuntimeError
if not version:
    raise RuntimeError
```
But because the code only saves the errors and raises them later all at once, static analysis wouldn't be able to prove that the types are correct otherwise.